File: csharp/queen-attack/QueenAttack.cs
public class Queen
{
    public Queen(int row, int column)
    {
        Row = row;
        Column = column;
    }

    public int Row { get; }
    public int Column { get; }
}

public static class QueenAttack
{
    public static bool CanAttack(Queen white, Queen black)
    {
        // I have no board, so I am assuming we have no other pieces
        // on the board that could block the queens.
        if ((white.Column == black.Column) && (white.Row == black.Row))
            throw new ArgumentException("Pieces cannot be at the same location.");


        // Horizontal and Vertical.
        if ((white.Row == black.Row) || (white.Column == black.Column))
            return true;

        // Diagonal.
        int dx = Math.Abs(white.Column - black.Column);
        int dy = Math.Abs(white.Row - black.Row);
        if (dx == dy)
            return true;

        return false;
    }

    public static Queen Create(int row, int column)
    {
        if ((column < 0) || (row < 0) || (column >= 8) || (row >= 8))
            throw new ArgumentOutOfRangeException("Queen is not on the board.");

        return new Queen(row, column);
    }
        
}

File: javascript/queen-attack/queen-attack.js
//
// This is only a SKELETON file for the 'Queen Attack' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class QueenAttack {
  constructor({
    black: [blackRow, blackColumn] = [0, 3],
    white: [whiteRow, whiteColumn] = [7, 3],
  } = {}) {

    if (blackRow === whiteRow && blackColumn === whiteColumn)
      throw new Error('Queens cannot share the same space')

    if (!this.onGameBoard(blackRow, blackColumn) || 
        !this.onGameBoard(whiteRow, whiteColumn))
      throw new Error('Queen must be placed on the board')
    
    this.black = [blackRow, blackColumn]
    this.white = [whiteRow, whiteColumn]
  }

  toString() {
    let ret = [
      ['_', '_', '_', '_', '_', '_', '_', '_']
    , ['_', '_', '_', '_', '_', '_', '_', '_']
    , ['_', '_', '_', '_', '_', '_', '_', '_']
    , ['_', '_', '_', '_', '_', '_', '_', '_']
    , ['_', '_', '_', '_', '_', '_', '_', '_']
    , ['_', '_', '_', '_', '_', '_', '_', '_']
    , ['_', '_', '_', '_', '_', '_', '_', '_']
    , ['_', '_', '_', '_', '_', '_', '_', '_']]

    ret[this.black[0]][this.black[1]] = 'B'
    ret[this.white[0]][this.white[1]] = 'W'

    return ret.map((row) => row.join(' ')).join('\n')
  }

  get canAttack() {
    let dx = Math.abs(this.black[1] - this.white[1])
    let dy = Math.abs(this.black[0] - this.white[0])
    return (dx === 0 || dy === 0 || dx === dy)
  }

  onGameBoard(y, x) {
    return x >= 0 && x < 8 && y >= 0 && y < 8
  }
  
}
